### PR TITLE
op-batcher: Multi-blob Support

### PR DESCRIPTION
**Pull request by sebastianst**
**Description**

Adds multi-blob support to the batcher.

**Tests**

Adapting all tests & adding a new multi-blob test.

See also https://github.com/ethereum-optimism/optimism/pull/9774 for PoC action test.

**Metadata**

- Fixes https://github.com/ethereum-optimism/protocol-quest/issues/113

### CODE REVIEW

File: op-batcher/batcher/driver.go [hunk 1]
```diff
@@ -413,15 +416,15 @@ func (l *BatchSubmitter) sendTransaction(ctx context.Context, txdata txData, que
...
-	l.Log.Info("building Blob transaction candidate", "size", len(data))
-	var b eth.Blob
-	if err := b.FromData(data); err != nil {
-		return nil, fmt.Errorf("data could not be converted to blob: %w", err)
-	}
+func (l *BatchSubmitter) blobTxCandidate(data txData) (*txmgr.TxCandidate, error) {
+	blobs := data.Blobs()
+	size := data.Len()
+	l.Log.Info("building Blob transaction candidate", "size", size, "num_blobs", len(blobs))
+	l.Metr.RecordBlobUsedBytes(size)
```
Comment (roberto-bayardo) [inline op-batcher/batcher/driver.go]: This metric was intended to allow us to observe how close we are to filling up single blobs.   Not sure if you're guaranteeing all but the last blob is 100% full in this CL yet (haven't reveived deeply) but if not we may need to tweak this?

Comment (sebastianst) [reply op-batcher/batcher/driver.go]: Yeah the metrics definitely need improving. We could measure three things
* total size of all blobs together (currently)
* size of last blob (the leftover)
* number of blobs in a tx

File: op-batcher/batcher/tx_data.go [hunk 2]
```diff
@@ -12,47 +14,100 @@ import (
...
+	)
+	for _, f := range id {
+		if f.chID == curChID {
+			sb.WriteString(fmt.Sprintf("+%d", f.frameNumber))
+		} else {
+			if curChID != (derive.ChannelID{}) {
+				sb.WriteString("|")
+			}
+			curChID = f.chID
+			sb.WriteString(fmt.Sprintf("%s:%d", f.chID.TerminalString(), f.frameNumber))
```
Comment (roberto-bayardo) [inline op-batcher/batcher/tx_data.go]: I guess this line is the only difference between TerminalString() and String()?  Perhaps remove the duplicate code by having a generic method that takes a functor to compute the chID string?

Comment (sebastianst) [reply op-batcher/batcher/tx_data.go]: I thought about that, but was too lazy at the time 😬  Will optimize this once everything else works.

File: op-e2e/eip4844_test.go [hunk 3]
```diff
@@ -108,11 +122,37 @@ func TestSystem4844E2E(t *testing.T) {
...
+	var numBlobs int
+	_, err = gethutils.FindBlock(l1Client, int(tip.Number.Int64()), 0, 5*time.Second,
+		func(b *types.Block) (bool, error) {
+			for _, tx := range b.Transactions() {
+				if tx.Type() != types.BlobTxType {
+					continue
+				}
+				numBlobs = len(tx.BlobHashes())
+				// expect to find at least one tx with more than 1 blob in multi-blob case
+				if !multiBlob || numBlobs > 1 {
```
Comment (roberto-bayardo) [inline op-e2e/eip4844_test.go]: is it worth also checking to make sure blobs other than the last one are fully utilized?

Comment (sebastianst) [reply op-e2e/eip4844_test.go]: Definitely something to test! But maybe we better do this in a unit test for the `ChannelBuilder`?

File: op-batcher/batcher/tx_data.go (hunk 2 above)
Comment (sebastianst) [reply op-batcher/batcher/tx_data.go]: done

File: op-e2e/eip4844_test.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-e2e/eip4844_test.go]: either way (or both) works for me.

Comment (sebastianst) [reply op-e2e/eip4844_test.go]: added these assertions to the test

File: op-batcher/batcher/driver.go (hunk 1 above)
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: I changed the existing metric to record the size of the last left-over blob. The number of blobs can already be accessed via the metrics tracking the number of frames, and the total channel size is also already captured by a metric. So I consider the metrics issue resolved for now.

File: op-batcher/batcher/channel_builder.go [hunk 4]
```diff
@@ -449,11 +464,13 @@ func (c *ChannelBuilder) NextFrame() frameData {
 	return f
 }
 
-// PushFrame adds the frame back to the internal frames queue. Panics if not of
+// PushFrames adds the frame back to the internal frames queue. Panics if not of
```
Comment (roberto-bayardo) [inline op-batcher/batcher/channel_builder.go]:
```suggestion
// PushFrames adds the frames back to the internal frames queue. Panics if not of
```

**Review by roberto-bayardo — APPROVED** [review]

**Review by trianglesphere — APPROVED** [review]

**Pull request merged**